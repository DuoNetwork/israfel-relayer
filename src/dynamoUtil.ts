import { SignedOrder } from '0x.js';
import AWS from 'aws-sdk';
import {
	AttributeMap,
	DeleteItemInput,
	PutItemInput,
	QueryInput,
	QueryOutput,
	ScanInput,
	ScanOutput,
	UpdateItemInput
} from 'aws-sdk/clients/dynamodb';
import * as CST from './constants';
import { IDuoSignedOrder, ILiveOrders, UserOrderOperation } from './types';
import util from './util';

class DynamoUtil {
	private ddb: undefined | AWS.DynamoDB = undefined;
	// private process: string = 'UNKNOWN';
	private live: boolean = false;
	private hostname: string = 'hostname';
	private tool: string = 'tool';
	// private contractUtil: ContractUtil | undefined = undefined;
	public init(
		config: object,
		live: boolean,
		tool: string
		//  process: string
	) {
		this.live = live;
		// this.process = process;
		this.tool = tool;
		AWS.config.update(config);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
		return Promise.resolve();
	}

	public  insertData(params: PutItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.putItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public queryData(params: QueryInput): Promise<QueryOutput> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.query(params, (err, data) => (err ? reject(err) : resolve(data)))
					: reject('dynamo db connection is not initialized')
		);
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.scan(params, (err, data) => (err ? reject(err) : resolve(data)))
					: reject('dynamo db connection is not initialized')
		);
	}

	public deleteData(params: DeleteItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.deleteItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public updateItem(params: UpdateItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.updateItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public convertDuoSignedOrderToDynamo(
		order: SignedOrder,
		orderHash: string,
		timestamp: number,
		side: string
	) {
		return {
			[CST.DB_ORDER_HASH]: { S: orderHash },
			[CST.DB_PRICE]: {
				N: order.makerAssetAmount.div(order.takerAssetAmount).valueOf() + ''
			},
			[CST.DB_FILLED_TAKER_ASSET_AMT]: { S: '0' },
			[CST.DB_REMAINING_MAKER_ASSET_AMT]: { S: order.makerAssetAmount.valueOf() + '' },
			[CST.DB_REMAINING_TAKER_ASSET_AMT]: { S: order.takerAssetAmount.valueOf() + '' },
			[CST.DB_SIDE]: { S: side },
			[CST.DB_UPDATED_AT]: { N: timestamp + '' }
		};
	}

	public async addLiveOrder(
		order: SignedOrder,
		orderHash: string,
		marketId: string,
		side: string
	) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS

		const data = this.convertDuoSignedOrderToDynamo(order, orderHash, systemTimestamp, side);

		const params = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_DEV}`,
			Item: {
				[CST.DB_PAIR]: {
					S: marketId
				},
				...data
			}
		};

		await this.insertData(params);
	}

	public async removeLiveOrder(pair: string, orderHash: string): Promise<void> {
		return this.deleteData({
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.${CST.DB_DEV}`,
			Key: {
				[CST.DB_PAIR]: {
					S: pair
				},
				[CST.DB_ORDER_HASH]: {
					S: orderHash
				}
			}
		});
	}

	public async deleteOrderSignature(orderHash: string): Promise<void> {
		return this.updateItem({
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.${CST.DB_DEV}`,
			Key: {
				[CST.DB_ORDER_HASH]: {
					S: orderHash
				}
			},
			ExpressionAttributeNames: {
				[CST.DB_SIGNATURE]: CST.DB_SIGNATURE
			},
			ExpressionAttributeValues: {
				[':' + CST.DB_SIGNATURE]: { S: '' }
			},
			UpdateExpression: `SET ${CST.DB_SIGNATURE} = ${':' + CST.DB_SIGNATURE}`
		});
	}

	public convertSignedOrderToDynamo(order: SignedOrder, timestamp: number) {
		return {
			[CST.DB_SENDER_ADDR]: { S: order.senderAddress + '' },
			[CST.DB_MAKER_ADDR]: { S: order.makerAddress + '' },
			[CST.DB_TAKER_ADDR]: { S: order.takerAddress + '' },
			[CST.DB_MAKER_FEE]: { S: order.makerFee.valueOf() + '' },
			[CST.DB_TAKER_FEE]: { S: order.takerFee.valueOf() + '' },
			[CST.DB_MAKER_ASSET_AMT]: { S: order.makerAssetAmount.valueOf() + '' },
			[CST.DB_TAKER_ASSET_AMT]: { S: order.takerAssetAmount.valueOf() + '' },
			[CST.DB_MAKER_ASSET_DATA]: { S: order.makerAssetData + '' },
			[CST.DB_TAKER_ASSET_DATA]: { S: order.takerAssetData + '' },
			[CST.DB_SALT]: { S: order.salt.valueOf() + '' },
			[CST.DB_EXCHANGE_ADDR]: { S: order.exchangeAddress + '' },
			[CST.DB_FEE_RECIPIENT_ADDR]: { S: order.feeRecipientAddress + '' },
			[CST.DB_EXPIRATION_TIME_SECONDS]: { S: order.expirationTimeSeconds.valueOf() + '' },
			[CST.DB_SIGNATURE]: { S: order.signature + '' },
			[CST.DB_UPDATED_AT]: { N: timestamp + '' }
		};
	}

	public async addRawOrder(order: SignedOrder, orderHash: string) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS
		const data = this.convertSignedOrderToDynamo(order, systemTimestamp);

		const params = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.live`
				: `${CST.DB_PROJECT}.${CST.DB_RAW_ORDERS}.dev`,
			Item: {
				[CST.DB_PAIR]: {
					S: orderHash
				},
				...data
			}
		};

		await this.insertData(params);
	}

	public async addUserOrder(
		account: string,
		orderHash: string,
		pair: string,
		operation: UserOrderOperation
	) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS

		const params = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_USER_ORDERS}.${CST.DB_LIVE}`
				: `${CST.DB_PROJECT}.${CST.DB_USER_ORDERS}.${CST.DB_DEV}`,
			Item: {
				[CST.DB_ACCOUNT]: {
					S: account
				},
				[CST.DB_PAIR_ORDERHASH]: {
					S: `${pair}|${orderHash}`
				},
				[CST.DB_OPERATION]: {
					S: operation
				},
				[CST.DB_UPDATED_AT]: { N: systemTimestamp + '' }
			}
		};

		await this.insertData(params);
	}

	public parseLiveOrders(data: AttributeMap): ILiveOrders {
		return {
			[CST.DB_PRICE]: Number(data[CST.DB_PRICE].S || '0'),
			[CST.DB_SIDE]: data[CST.DB_SIDE].S || '',
			[CST.DB_AMT]: Number(data[CST.DB_MAKER_ASSET_AMT].S || '0')
		};
	}

	public async getLiveOrders(pair: string): Promise<ILiveOrders[]> {
		const params: QueryInput = {
			TableName: this.live
				? `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.live`
				: `${CST.DB_PROJECT}.${CST.DB_LIVE_ORDERS}.dev`,
			KeyConditionExpression: `${CST.DB_PAIR} = :${CST.DB_PAIR}`,
			ExpressionAttributeValues: {
				[':' + CST.DB_PAIR]: { S: pair }
			}
		};

		const data = await this.queryData(params);
		if (!data.Items || !data.Items.length) return [];

		const parsedLiveOrders = data.Items.map(ob => this.parseLiveOrders(ob)).sort(
			(a, b) => b.price - a.price
		);

		return parsedLiveOrders;
	}

	public updateStatus(process: string) {
		const update: AttributeMap = {
			[CST.DB_STS_PROCESS]: {
				S: `${this.tool}|${process}|${this.hostname}`
			},
			[CST.DB_STS_HOSTNAME]: {
				S: this.hostname
			},
			[CST.DB_UPDATED_AT]: { N: util.getUTCNowTimestamp() + '' }
		};

		return this.insertData({
			TableName: `${CST.DB_PROJECT}.${CST.DB_STATUS}.${this.live ? CST.DB_LIVE : CST.DB_DEV}`,
			Item: update
		}).catch(error => {
			util.logError('Error insert status: ' + error);
			return false;
		});
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
