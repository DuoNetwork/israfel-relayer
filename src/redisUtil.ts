// import Redis from 'ioredis';
// import * as CST from '../common/constants';
// import { IOrderBook, ITrade } from '../common/types';
// import util from './util';

// export class RedisUtil {
// 	private redisPub: Redis.Redis | null = null;
// 	private redisSub: Redis.Redis | null = null;
// 	private handleOrderBooksUpdate:
// 		| ((channel: string, orderBooks: { [src: string]: IOrderBook }) => any)
// 		| null = null;
// 	private handleTradesUpdate: ((channel: string, trades: ITrade[]) => any) | null = null;

// 	public init(redisKey: { host: string; password: string; servername: string }) {
// 		this.redisPub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
// 		this.redisSub.on('pmessage', (pattern, channel, message) =>
// 			this.onMessage(channel, message, pattern)
// 		);
// 	}

// 	public onMessage(channel: string, message: string, pattern: string = '') {
// 		util.logDebug(pattern + channel);
// 		const [type, ...others] = channel.split('|');
// 		switch (type) {
// 			case CST.DB_ORDER_BOOKS:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, { [others[0]]: JSON.parse(message) });
// 				break;
// 			case CST.DB_OB_OVERVIEW:
// 			case CST.DB_OB_GLOBAL:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, JSON.parse(message));
// 				break;
// 			case CST.DB_TRADES:
// 				if (this.handleTradesUpdate) this.handleTradesUpdate(channel, JSON.parse(message));
// 				break;
// 			default:
// 				break;
// 		}
// 	}

// 	public onOrderBooks(
// 		handleOrderBooksUpdate: (channel: string, orderBooks: { [src: string]: IOrderBook }) => any
// 	) {
// 		this.handleOrderBooksUpdate = handleOrderBooksUpdate;
// 	}

// 	public onTrades(handleTradesUpdate: (channel: string, trades: ITrade[]) => any) {
// 		this.handleTradesUpdate = handleTradesUpdate;
// 	}

// 	public publish(channel: string, msg: string): void {
// 		if (this.redisPub) this.redisPub.publish(channel, msg);
// 	}

// 	public subscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.subscribe(channel);
// 	}

// 	public patternSubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.psubscribe(pattern);
// 	}

// 	public unsubscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.unsubscribe(channel);
// 	}

// 	public patternUnsubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.punsubscribe(pattern);
// 	}

// 	public push(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.lpush(key, values);
// 	}

// 	public putBack(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.rpush(key, values);
// 	}

// 	public async pop(key: string): Promise<string> {
// 		if (this.redisSub) return await this.redisSub.rpop(key);
// 		return '';
// 	}

// 	public async getQueueLength(key: string): Promise<number> {
// 		if (this.redisSub) return this.redisSub.llen(key);
// 		return 0;
// 	}
// }

// const redisUtil = new RedisUtil();
// export default redisUtil;
