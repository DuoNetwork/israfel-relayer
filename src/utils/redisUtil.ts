import Redis from 'ioredis';
import * as CST from '../common/constants';
import {  IOrderBookSnapshotUpdate, IOrderQueueItem } from '../common/types';
import util from './util';

class RedisUtil {
	private redisPub: Redis.Redis | null = null;
	private redisSub: Redis.Redis | null = null;
	private handleOrderBookUpdate:
		| ((channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any)
		| null = null;
	private handleOrderUpdate:
		| ((channel: string, orderQueueItem: IOrderQueueItem) => any)
		| null = null;

	public init(redisKey: { host: string; password: string; servername: string }) {
		this.redisPub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
		this.redisSub.on('pmessage', (pattern, channel, message) =>
			this.onMessage(channel, message, pattern)
		);
	}

	public onMessage(channel: string, message: string, pattern: string = '') {
		util.logDebug(pattern + channel + message);
		const type = channel.split('|')[0];
		switch (type) {
			case CST.DB_ORDER_BOOKS:
				if (this.handleOrderBookUpdate)
					this.handleOrderBookUpdate(channel, JSON.parse(message));
				break;
			case CST.DB_ORDERS:
				if (this.handleOrderUpdate) this.handleOrderUpdate(channel, JSON.parse(message));
				break;
			default:
				break;
		}
	}

	public onOrderBookUpdate(
		handleOrderBookUpdate: (channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any
	) {
		this.handleOrderBookUpdate = handleOrderBookUpdate;
	}
	public onOrderUpdate(
		handleOrderUpdate: (channel: string, orderQueueItem: IOrderQueueItem) => any
	) {
		this.handleOrderUpdate = handleOrderUpdate;
	}

	public publish(channel: string, msg: string) {
		if (this.redisPub) return this.redisPub.publish(channel, msg);
		return Promise.resolve(0);
	}

	public increment(key: string) {
		if (this.redisPub) return this.redisPub.incr(key);
		return Promise.reject();
	}

	public set(key: string, value: string) {
		if (this.redisPub) return this.redisPub.set(key, value);
		return Promise.resolve('');
	}

	public get(key: string) {
		if (this.redisPub) return this.redisPub.get(key);
		return Promise.resolve('');
	}

	public hashSet(key: string, field: string, value: string) {
		if (this.redisPub) return this.redisPub.hset(key, field, value);
		return Promise.resolve(0);
	}

	public hashGet(key: string, field: string): Promise<string | null> {
		if (this.redisPub) return this.redisPub.hget(key, field);
		return Promise.resolve(null);
	}

	public async hashMultiGet(key: string, ...fields: string[]) {
		if (this.redisPub) {
			const values: Array<string | null> = await this.redisPub.hmget(key, ...fields);
			const output: { [field: string]: string | null } = {};
			fields.forEach((f, i) => (output[f] = values[i]));
			return output;
		}
		return Promise.resolve({});
	}

	public hashGetAll(key: string): Promise<any> {
		if (this.redisPub) return this.redisPub.hgetall(key);
		return Promise.resolve(null);
	}

	public hashDelete(key: string, field: string) {
		if (this.redisPub) return this.redisPub.hdel(key, field);
	}

	public subscribe(channel: string) {
		if (this.redisSub) this.redisSub.subscribe(channel);
	}

	public patternSubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.psubscribe(pattern);
	}

	public unsubscribe(channel: string) {
		if (this.redisSub) this.redisSub.unsubscribe(channel);
	}

	public patternUnsubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.punsubscribe(pattern);
	}

	public push(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.lpush(key, values);
	}

	public putBack(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.rpush(key, values);
	}

	public pop(key: string) {
		if (this.redisPub) return this.redisPub.rpop(key);
		return Promise.resolve('');
	}

	public multi() {
		if (this.redisPub) return this.redisPub.multi({ pipeline: false });
		return Promise.resolve('');
	}

	public exec() {
		if (this.redisPub) return this.redisPub.exec();
		return Promise.resolve('');
	}

	public getQueueLength(key: string) {
		if (this.redisPub) return this.redisPub.llen(key);
		return Promise.resolve(0);
	}
}

const redisUtil = new RedisUtil();
export default redisUtil;
