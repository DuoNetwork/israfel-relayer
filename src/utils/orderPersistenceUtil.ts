import { SignedOrder } from '0x.js';
import * as CST from '../common/constants';
import {
	ILiveOrder,
	IOption,
	IOrderPersistRequest,
	IOrderQueueItem,
	IStringSignedOrder,
	IUserOrder
} from '../common/types';
import dynamoUtil from './dynamoUtil';
import redisUtil from './redisUtil';
import util from './util';
import Web3Util from './Web3Util';

class OrderPersistenceUtil {
	public subscribeOrderUpdate(
		pair: string,
		handleOrderUpdate: (channel: string, orderQueueItem: IOrderQueueItem) => any
	) {
		redisUtil.onOrderUpdate(handleOrderUpdate);
		redisUtil.subscribe(`${CST.DB_ORDERS}|${CST.DB_PUBSUB}|${pair}`);
	}

	public unsubscribeOrderUpdate(pair: string) {
		redisUtil.unsubscribe(`${CST.DB_ORDERS}|${CST.DB_PUBSUB}|${pair}`);
	}

	public async addUserOrderToDB(
		liveOrder: ILiveOrder,
		type: string,
		status: string,
		updatedBy: string,
		processed: boolean
	) {
		const userOrder = this.constructUserOrder(liveOrder, type, status, updatedBy, processed);
		try {
			await dynamoUtil.addUserOrder(userOrder);
			util.logDebug(`added user order ${liveOrder.orderHash}|${type}|${status}|${updatedBy}`);
		} catch (error) {
			util.logError(error);
		}

		return userOrder;
	}

	public async getLiveOrderInPersistence(pair: string, orderHash: string) {
		const queueStrings = await redisUtil.hashMultiGet(
			`${CST.DB_ORDERS}|${CST.DB_CACHE}`,
			`${CST.DB_TERMINATE}|${orderHash}`,
			`${CST.DB_UPDATE}|${orderHash}`,
			`${CST.DB_ADD}|${orderHash}`
		);
		if (queueStrings[`${CST.DB_TERMINATE}|${orderHash}`]) return null;

		const updateQueueString = queueStrings[`${CST.DB_UPDATE}|${orderHash}`];
		if (updateQueueString) {
			const orderQueueItem: IOrderQueueItem = JSON.parse(updateQueueString);
			return orderQueueItem.liveOrder;
		}

		const addQueueString = queueStrings[`${CST.DB_ADD}|${orderHash}`];
		if (addQueueString) {
			const orderQueueItem: IOrderQueueItem = JSON.parse(addQueueString);
			return orderQueueItem.liveOrder;
		}

		const liveOrders = await dynamoUtil.getLiveOrders(pair, orderHash);
		if (liveOrders.length < 1) return null;

		return liveOrders[0];
	}

	public async getAllLiveOrdersInPersistence(pair: string) {
		const redisOrders = await redisUtil.hashGetAll(`${CST.DB_ORDERS}|${CST.DB_CACHE}`);
		const dynamoOrders = await dynamoUtil.getLiveOrders(pair);

		const addOrders: { [orderHash: string]: ILiveOrder } = {};
		const updateOrders: { [orderHash: string]: ILiveOrder } = {};
		const terminateOrders: { [orderHash: string]: ILiveOrder } = {};
		for (const key in redisOrders) {
			const [method, orderHash] = key.split('|');
			const orderQueueItem: IOrderQueueItem = JSON.parse(redisOrders[key]);
			if (method === CST.DB_TERMINATE) terminateOrders[orderHash] = orderQueueItem.liveOrder;
			else if (method === CST.DB_ADD) addOrders[orderHash] = orderQueueItem.liveOrder;
			else updateOrders[orderHash] = orderQueueItem.liveOrder;
		}

		const allOrders: { [orderHash: string]: ILiveOrder } = {};
		dynamoOrders.forEach(o => (allOrders[o.orderHash] = o));
		Object.assign(allOrders, addOrders);
		Object.assign(allOrders, updateOrders);
		for (const orderHash in terminateOrders)
			if (allOrders[orderHash]) delete allOrders[orderHash];

		return allOrders;
	}

	public async getRawOrderInPersistence(orderHash: string) {
		const redisStringOrder = await redisUtil.hashGet(
			`${CST.DB_ORDERS}|${CST.DB_CACHE}`,
			`${CST.DB_ADD}|${orderHash}`
		);
		if (redisStringOrder) {
			const redisRawOrder: IOrderQueueItem = JSON.parse(redisStringOrder);
			return {
				orderHash: orderHash,
				signedOrder: redisRawOrder.signedOrder as IStringSignedOrder
			};
		}

		const dynamoRawOrder = await dynamoUtil.getRawOrder(orderHash);
		return dynamoRawOrder;
	}

	public async persistOrder(orderPersistRequest: IOrderPersistRequest) {
		const {
			pair,
			orderHash,
			method,
			balance,
			side,
			fill,
			status,
			requestor
		} = orderPersistRequest;
		if (method === CST.DB_ADD && !side) {
			util.logDebug(`invalid add request ${orderHash}, missing side`);
			return null;
		}

		let liveOrder = await this.getLiveOrderInPersistence(pair, orderHash);
		if (method === CST.DB_ADD && liveOrder) {
			util.logDebug(`order ${orderHash} already exist, ignore add request`);
			return null;
		} else if (method !== CST.DB_ADD && !liveOrder) {
			util.logDebug(`order ${orderHash} does not exist, ignore ${method} request`);
			return null;
		}

		const sequence = await redisUtil.increment(`${CST.DB_SEQUENCE}|${pair}`);
		if (method === CST.DB_ADD) {
			liveOrder = this.constructNewLiveOrder(
				orderPersistRequest.signedOrder as IStringSignedOrder,
				pair,
				side || '',
				orderHash
			);
			liveOrder.initialSequence = sequence;
		}
		const orderQueueItem: IOrderQueueItem = {
			method: method,
			status: status,
			requestor: requestor,
			liveOrder: liveOrder as ILiveOrder
		};
		orderQueueItem.liveOrder.currentSequence = sequence;
		if (method === CST.DB_ADD) orderQueueItem.signedOrder = orderPersistRequest.signedOrder;
		else if (orderPersistRequest.status === CST.DB_FILL) {
			orderQueueItem.liveOrder.fill = orderQueueItem.liveOrder.amount;
			orderQueueItem.liveOrder.balance = 0;
		} else {
			if (fill) orderQueueItem.liveOrder.fill = fill;
			if (balance !== -1) orderQueueItem.liveOrder.balance = balance;
		}

		util.logDebug(`storing order queue item in redis ${orderHash}`);
		await redisUtil.multi();
		const key = `${method}|${orderHash}`;
		// store order in hash map
		redisUtil.hashSet(`${CST.DB_ORDERS}|${CST.DB_CACHE}`, key, JSON.stringify(orderQueueItem));
		// push orderhash into queue
		redisUtil.push(`${CST.DB_ORDERS}|${CST.DB_QUEUE}`, key);
		await redisUtil.exec();
		util.logDebug(`done`);

		try {
			redisUtil.publish(
				`${CST.DB_ORDERS}|${CST.DB_PUBSUB}|${pair}`,
				JSON.stringify(orderQueueItem)
			);
		} catch (error) {
			util.logError(error);
		}

		return this.addUserOrderToDB(orderQueueItem.liveOrder, method, status, requestor, false);
	}

	public constructUserOrder(
		liveOrder: ILiveOrder,
		type: string,
		status: string,
		updatedBy: string,
		processed: boolean
	): IUserOrder {
		return {
			...liveOrder,
			type: type,
			status: status,
			updatedBy: updatedBy,
			processed: processed
		};
	}

	public constructNewLiveOrder(
		signedOrder: IStringSignedOrder,
		pair: string,
		side: string,
		orderHash: string
	): ILiveOrder {
		const isBid = side === CST.DB_BID;
		const amount = Web3Util.fromWei(
			isBid ? signedOrder.takerAssetAmount : signedOrder.makerAssetAmount
		);
		return {
			account: signedOrder.makerAddress,
			pair: pair,
			orderHash: orderHash,
			price: Web3Util.getPriceFromSignedOrder(signedOrder, side),
			amount: amount,
			balance: amount,
			fill: 0,
			side: side,
			expiry: Number(signedOrder.expirationTimeSeconds) * 1000,
			initialSequence: 0,
			currentSequence: 0,
			createdAt: util.getUTCNowTimestamp()
		};
	}

	public async processOrderQueue() {
		const queueKey = await redisUtil.pop(`${CST.DB_ORDERS}|${CST.DB_QUEUE}`);
		if (!queueKey) return false;

		const queueItemString = await redisUtil.hashGet(
			`${CST.DB_ORDERS}|${CST.DB_CACHE}`,
			queueKey
		);
		util.logDebug(`processing order: ${queueKey}`);
		if (!queueItemString) {
			util.logDebug('empty queue item, ignore');
			return true;
		}
		const [method, orderHash] = queueKey.split('|');
		const orderQueueItem: IOrderQueueItem = JSON.parse(queueItemString);
		try {
			util.logDebug(`${method} order`);
			if (method === CST.DB_ADD) {
				await dynamoUtil.addRawOrder({
					orderHash: orderHash,
					signedOrder: orderQueueItem.signedOrder as IStringSignedOrder
				});
				util.logDebug(`added raw order`);
				await dynamoUtil.addLiveOrder(orderQueueItem.liveOrder);
				util.logDebug(`added live order`);
			} else if (method === CST.DB_TERMINATE) {
				await dynamoUtil.deleteRawOrderSignature(orderHash);
				util.logDebug(`deleted raw order`);
				await dynamoUtil.deleteLiveOrder(orderQueueItem.liveOrder);
				util.logDebug(`deleted live order`);
			} else {
				await dynamoUtil.updateLiveOrder(orderQueueItem.liveOrder);
				util.logDebug(`added live order`);
			}
			redisUtil.hashDelete(`${CST.DB_ORDERS}|${CST.DB_CACHE}`, queueKey);
			util.logDebug(`removed redis data`);
		} catch (err) {
			util.logError(`error in processing for ${queueKey}`);
			util.logError(err);
			await redisUtil.multi();
			redisUtil.hashSet(`${CST.DB_ORDERS}|${CST.DB_CACHE}`, queueKey, queueItemString);
			redisUtil.putBack(`${CST.DB_ORDERS}|${CST.DB_QUEUE}`, queueKey);
			await redisUtil.exec();
			return false;
		}

		await this.addUserOrderToDB(
			orderQueueItem.liveOrder,
			method,
			orderQueueItem.status,
			orderQueueItem.requestor,
			true
		);

		return true;
	}

	public parseSignedOrder(order: IStringSignedOrder): SignedOrder {
		const {
			makerFee,
			takerFee,
			makerAssetAmount,
			takerAssetAmount,
			salt,
			expirationTimeSeconds,
			...rest
		} = order;
		return {
			...rest,
			makerFee: Web3Util.stringToBN(makerFee),
			takerFee: Web3Util.stringToBN(takerFee),
			makerAssetAmount: Web3Util.stringToBN(makerAssetAmount),
			takerAssetAmount: Web3Util.stringToBN(takerAssetAmount),
			salt: Web3Util.stringToBN(salt),
			expirationTimeSeconds: Web3Util.stringToBN(expirationTimeSeconds)
		};
	}

	public async startProcessing(option: IOption) {
		if (option.server) {
			dynamoUtil.updateStatus(
				CST.DB_ORDERS,
				await redisUtil.getQueueLength(`${CST.DB_ORDERS}|${CST.DB_QUEUE}`)
			);

			setInterval(
				async () =>
					dynamoUtil.updateStatus(
						CST.DB_ORDERS,
						await redisUtil.getQueueLength(`${CST.DB_ORDERS}|${CST.DB_QUEUE}`)
					),
				15000
			);
		}

		const loop = () =>
			this.processOrderQueue().then(result => {
				setTimeout(() => loop(), result ? 0 : 500);
			});
		loop();
	}
}
const orderPersistenceUtil = new OrderPersistenceUtil();
export default orderPersistenceUtil;
