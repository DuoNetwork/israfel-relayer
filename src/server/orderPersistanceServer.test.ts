import dynamoUtil from '../utils/dynamoUtil';
import orderPersistenceUtil from '../utils/orderPersistenceUtil';
import redisUtil from '../utils/redisUtil';
import orderPersistenceServer from './orderPersistanceServer';

test('startProcessing, server', async () => {
	dynamoUtil.updateStatus = jest.fn();
	redisUtil.getQueueLength = jest.fn(() => Promise.resolve(10));
	global.setInterval = jest.fn();
	global.setTimeout = jest.fn();

	let result = false;
	orderPersistenceUtil.processOrderQueue = jest.fn(() => Promise.resolve(result));
	await orderPersistenceServer.startProcessing({ server: true } as any);

	expect((global.setInterval as jest.Mock).mock.calls).toMatchSnapshot();

	await (global.setInterval as jest.Mock).mock.calls[0][0]();
	expect((redisUtil.getQueueLength as jest.Mock).mock.calls).toMatchSnapshot();
	expect((dynamoUtil.updateStatus as jest.Mock).mock.calls).toMatchSnapshot();
	result = true;

	await (global.setTimeout as jest.Mock).mock.calls[0][0]();
	expect((global.setTimeout as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startProcessing, no server', async () => {
	dynamoUtil.updateStatus = jest.fn();
	redisUtil.getQueueLength = jest.fn(() => Promise.resolve(10));
	global.setInterval = jest.fn();
	global.setTimeout = jest.fn();

	await orderPersistenceServer.startProcessing({ server: false } as any);

	expect(dynamoUtil.updateStatus as jest.Mock).not.toBeCalled();
	expect(redisUtil.getQueueLength as jest.Mock).not.toBeCalled();
	expect(global.setInterval as jest.Mock).not.toBeCalled();
});
