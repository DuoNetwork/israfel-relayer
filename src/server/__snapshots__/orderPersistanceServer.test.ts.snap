// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`startProcessing, server 1`] = `
Array [
  Array [
    [Function],
    15000,
  ],
]
`;

exports[`startProcessing, server 2`] = `
Array [
  Array [
    "orders|queue",
  ],
  Array [
    "orders|queue",
  ],
]
`;

exports[`startProcessing, server 3`] = `
Array [
  Array [
    "orders",
    10,
  ],
  Array [
    "orders",
    10,
  ],
]
`;

exports[`startProcessing, server 4`] = `
Array [
  Array [
    [Function],
    500,
  ],
  Array [
    [Function],
    0,
  ],
]
`;
